src/scanner.c

```diff
@@ -0,0 +1,200 @@
+
+#define MAX(a, b) ((a) > (b) ? (a) : (b))
+
+#define STRING_RESIZE(vec, _cap)                                             \
+    void *block = realloc((vec).data, ((_cap) + 1) * sizeof((vec).data[0])); \
+    assert(block != NULL);                                                   \
+    (vec).data = block;                                                      \
+    memset((vec).data + (vec).len,                                           \
+           0,                                                                \
+           (((_cap) + 1) - (vec).len) * sizeof((vec).data[0]));              \
+    (vec).cap = (_cap);
+
+#define STRING_GROW(vec, _cap)        \
+    if ((vec).cap < (_cap)) {         \
+        STRING_RESIZE((vec), (_cap)); \
+    }
+
+#define STRING_PUSH(vec, el)                          \
+    if ((vec).cap == (vec).len) {                     \
+        STRING_RESIZE((vec), MAX(16, (vec).len * 2)); \
+    }                                                 \
+    (vec).data[(vec).len++] = (el);
+
+#define STRING_FREE(vec)        \
+    {                           \
+        if ((vec).data != NULL) \
+            free((vec).data);   \
+    }
+
+#define STRING_CLEAR(vec)                                \
+    {                                                    \
+        (vec).len = 0;                                   \
+        memset((vec).data, 0, (vec).cap * sizeof(char)); \
+    }
+
+enum TokenType {
+    HERE_STRING_BODY,
+};
+
+typedef struct {
+    uint32_t len;
+    uint32_t cap;
+    char *data;
+} String;
+
+String string_new() {
+    return (String){.cap = 16, .len = 0, .data = calloc(1, sizeof(char) * 17)};
+}
+
+typedef struct {
+    bool valid;
+    String terminator_str;
+    String current_line;
+} Scanner;
+
+static void advance(TSLexer *lexer) {
+    lexer->advance(lexer, false);
+}
+
+// NOTE: only "\n" is allowed as newline here,
+// It implies that "\r" can also be terminator.
+static bool isnewline(int32_t chr) {
+    return chr == '\n';
+}
+
+static void read_terminator(Scanner *scanner, TSLexer *lexer) {
+    while (true) {
+        if (isnewline(lexer->lookahead)) {
+            return;
+        }
+        if (lexer->eof(lexer)) {
+            scanner->valid = false;
+            return;
+        }
+        STRING_PUSH(scanner->terminator_str, lexer->lookahead);
+        advance(lexer);
+    }
+}
+
+// `read_line` read strings until a newline or EOF
+static void read_line(Scanner *scanner, TSLexer *lexer) {
+    while (!isnewline(lexer->lookahead) && !lexer->eof(lexer)) {
+        STRING_PUSH(scanner->current_line, lexer->lookahead);
+        advance(lexer);
+    }
+}
+
+// Suppose terminator is `T`, newline (\n) is `$`,
+// It should accept "#<<T$T" or "#<<T$...$T", where `...`
+// is the string content.
+bool scan(Scanner *scanner, TSLexer *lexer, const bool *valid_symbols) {
+    if (!valid_symbols[HERE_STRING_BODY]) {
+        return false;
+    }
+
+    STRING_CLEAR(scanner->terminator_str);
```

Comment:
Consider allocate and free `scanner` struct here, and keep `tree_sitter_racket_external_scanner_*` functions empty.

Reply:
Also consider rename `scanner` to `here_string_states` or something like that.

Reply:
I disagree because serialize/deserialize are called multiple times in between scans.

Scanner is also the common naming convention in any ts repo

Reply:
What the `scanner` means here? We know the whole file is for scanner. The name doesn't reveal useful information.

The struct

```
typedef struct {
    bool valid;
    String terminator_str;
    String current_line;
} Scanner;
```

stores states that relates to here string parse not general scanner states.
I think `here_string_states` is a more precise name.

Reply:
> What the scanner means here? We know the whole file is for scanner. The name doesn't reveal useful information.

The memory allocated by the external scanner (scanner.c) for all the data/information it contains, that I (and every other ts-repo) find fitting. I think the field names give you the information needed for what's going on..but it is your repo and I can change it if you'd like